File: packages/oidc/android/src/main/java/com/pingidentity/rnoidc/OidcClientFactory.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rnoidc

import com.pingidentity.oidc.OidcClient
import com.pingidentity.oidc.OidcClientConfig
import com.pingidentity.oidc.OidcWebClient
import com.pingidentity.oidc.OpenIdConfiguration
import com.pingidentity.oidc.module.Oidc
import com.pingidentity.storage.CacheStrategy
import com.pingidentity.storage.EncryptedDataStoreStorageConfig
import com.pingidentity.rncore.registry.Registry
import com.pingidentity.rncore.storage.StorageConfigHandleContract

/**
 * Builds native OIDC client instances from JS payloads.
 */
internal class OidcClientFactory(
  private val storageRegistry: Registry,
  /** Resolver that maps logger ids to native logger instances. */
  private val loggerResolver: (String?) -> com.pingidentity.logger.Logger?
) {

  /**
   * Build a native OIDC config for a device client.
   *
   * @param config Parsed JS client payload
   * @return Configured native OIDC config
   */
  fun buildOidcClientConfig(config: OidcClientPayload): OidcClientConfig {
    val resolvedLogger = loggerResolver(config.loggerId)
    return OidcClientConfig().apply {
      resolvedLogger?.let { logger = it }
      config.discoveryEndpoint?.let { discoveryEndpoint = it }
      clientId = config.clientId
      redirectUri = config.redirectUri
      scopes = config.scopes.toMutableSet()
      acrValues = config.acrValues
      signOutRedirectUri = config.signOutRedirectUri
      state = config.state
      nonce = config.nonce
      uiLocales = config.uiLocales
      config.refreshThreshold?.let { refreshThreshold = it }
      loginHint = config.loginHint
      display = config.display
      prompt = config.prompt
      if (config.additionalParameters.isNotEmpty()) {
        additionalParameters = config.additionalParameters
      }
      applyOpenIdIfPresent(config.openId, config.discoveryEndpoint)
      applyStorageIfPresent(config.storageId)
    }
  }

  /**
   * Build a native OIDC web client from stored JS configuration.
   *
   * @param config Parsed JS client payload
   * @return Configured native OIDC web client
   */
  fun buildWebClient(config: OidcClientPayload): OidcWebClient {
    val resolvedLogger = loggerResolver(config.loggerId)
    return OidcWebClient {
      module(Oidc) {
        resolvedLogger?.let { logger = it }
        config.discoveryEndpoint?.let { discoveryEndpoint = it }
        clientId = config.clientId
        redirectUri = config.redirectUri
        scopes = config.scopes.toMutableSet()
        acrValues = config.acrValues
        signOutRedirectUri = config.signOutRedirectUri
        state = config.state
        nonce = config.nonce
        uiLocales = config.uiLocales
        config.refreshThreshold?.let { refreshThreshold = it }
        loginHint = config.loginHint
        display = config.display
        prompt = config.prompt
        if (config.additionalParameters.isNotEmpty()) {
          additionalParameters = config.additionalParameters
        }
        applyOpenIdIfPresent(config.openId, config.discoveryEndpoint)
        applyStorageIfPresent(config.storageId)
      }
    }
  }

  /**
   * Build a native OIDC client instance.
   *
   * @param config Parsed JS client payload
   * @return Configured native OIDC client
   */
  fun buildOidcClient(config: OidcClientPayload): OidcClient {
    val resolvedLogger = loggerResolver(config.loggerId)
    return OidcClient {
      resolvedLogger?.let { logger = it }
      config.discoveryEndpoint?.let { discoveryEndpoint = it }
      clientId = config.clientId
      redirectUri = config.redirectUri
      scopes = config.scopes.toMutableSet()
      acrValues = config.acrValues
      signOutRedirectUri = config.signOutRedirectUri
      state = config.state
      nonce = config.nonce
      uiLocales = config.uiLocales
      config.refreshThreshold?.let { refreshThreshold = it }
      loginHint = config.loginHint
      display = config.display
      prompt = config.prompt
      if (config.additionalParameters.isNotEmpty()) {
        additionalParameters = config.additionalParameters
      }
      applyOpenIdIfPresent(config.openId, config.discoveryEndpoint)
      applyStorageIfPresent(config.storageId)
    }
  }

  /**
   * Apply storage configuration from Core registry if an id is provided.
   *
   * @param storageId Storage configuration identifier
   */
  private fun OidcClientConfig.applyStorageIfPresent(storageId: String?) {
    if (storageId.isNullOrBlank()) {
      return
    }
    val storageConfig =
      storageRegistry.resolve(storageId) as? StorageConfigHandleContract
        ?: throw IllegalArgumentException("No storage config registered for id=$storageId")
    storage {
      applyStorageConfig(storageConfig)
    }
  }

  /**
   * Apply OpenID configuration fields if an explicit payload is provided.
   *
   * When a discovery endpoint is also configured, the payload is applied as
   * an override on top of the discovered configuration via
   * [OidcClientConfig.openIdOverride] -- only the fields present in the
   * payload replace their discovered value, so an override that supplies
   * only [OpenIdPayload.deviceAuthorizationEndpoint] (for a provider whose
   * discovery document omits it) does not blank out the other endpoints
   * discovery would otherwise resolve.
   *
   * When no discovery endpoint is configured, there is nothing to merge
   * onto: the payload is applied directly as the complete configuration,
   * matching [OpenIdConfiguration]'s own blank-string defaults for any
   * field the payload leaves unset.
   *
   * @param openId Optional OpenID configuration payload
   * @param discoveryEndpoint Discovery endpoint from the same JS payload, if any
   */
  private fun OidcClientConfig.applyOpenIdIfPresent(
    openId: OpenIdPayload?,
    discoveryEndpoint: String?
  ) {
    if (openId == null) {
      return
    }
    if (discoveryEndpoint.isNullOrBlank()) {
      this.openId = OpenIdConfiguration(
        authorizationEndpoint = openId.authorizationEndpoint ?: "",
        tokenEndpoint = openId.tokenEndpoint ?: "",
        userinfoEndpoint = openId.userinfoEndpoint ?: "",
        endSessionEndpoint = openId.endSessionEndpoint ?: "",
        pingEndIdpSessionEndpoint = openId.pingEndIdpSessionEndpoint ?: "",
        revocationEndpoint = openId.revocationEndpoint ?: "",
        deviceAuthorizationEndpoint = openId.deviceAuthorizationEndpoint ?: ""
      )
      return
    }
    openIdOverride = {
      openId.authorizationEndpoint?.let { authorizationEndpoint = it }
      openId.tokenEndpoint?.let { tokenEndpoint = it }
      openId.userinfoEndpoint?.let { userinfoEndpoint = it }
      openId.endSessionEndpoint?.let { endSessionEndpoint = it }
      openId.pingEndIdpSessionEndpoint?.let { pingEndIdpSessionEndpoint = it }
      openId.revocationEndpoint?.let { revocationEndpoint = it }
      openId.deviceAuthorizationEndpoint?.let { deviceAuthorizationEndpoint = it }
    }
  }

  /**
   * Apply storage configuration overrides to the encrypted data store.
   *
   * @param config Parsed JS storage configuration
   */
  private fun EncryptedDataStoreStorageConfig.applyStorageConfig(config: StorageConfigHandleContract) {
    config.fileName?.let { fileName = it }
    config.keyAlias?.let { keyAlias = it }
    config.strongBoxPreferred?.let { strongBoxPreferred = it }
    config.cacheStrategy?.let { cacheStrategy = parseCacheStrategy(it) }
  }

  /**
   * Parse cache strategy strings into native CacheStrategy values.
   *
   * @param rawValue Raw strategy name from JS
   * @return Parsed cache strategy, defaulting to NO_CACHE on unknown values
   */
  private fun parseCacheStrategy(rawValue: String): CacheStrategy {
    return when (rawValue.lowercase()) {
      "cache_on_failure" -> CacheStrategy.CACHE_ON_FAILURE
      "no_cache" -> CacheStrategy.NO_CACHE
      "cache" -> CacheStrategy.CACHE
      else -> runCatching { CacheStrategy.valueOf(rawValue) }.getOrDefault(CacheStrategy.NO_CACHE)
    }
  }
}

File: packages/oidc/ios/OidcClientFactory.swift
//
//  OidcClientFactory.swift
//  RNPingOidc
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import PingLogger
import PingOidc
import PingStorage
import PingBrowser
import RNPingCore

/// Builder helpers for OIDC native clients.
enum OidcClientFactory {
  /// Build a native OIDC client configuration from the payload.
  ///
  /// - Parameter payload: Parsed JS client configuration payload.
  /// - Parameter logger: Optional native logger resolved from Core registry.
  /// - Returns: Native OIDC client configuration.
  ///
  /// Naming matches the Android factory for parity.
  static func buildOidcClient(
    _ payload: OidcClientPayload,
    logger: Logger?,
    queueKey: DispatchSpecificKey<Void>? = nil
  ) -> OidcClientConfig {
    let config = OidcClientConfig()
    config.clientId = payload.clientId
    config.discoveryEndpoint = payload.discoveryEndpoint ?? ""
    config.redirectUri = payload.redirectUri
    config.scopes = Set(payload.scopes)
    config.acrValues = payload.acrValues
    config.state = payload.state
    config.nonce = payload.nonce
    config.uiLocales = payload.uiLocales
    config.loginHint = payload.loginHint
    config.display = payload.display
    config.prompt = payload.prompt
    if let refreshThreshold = payload.refreshThreshold {
      config.refreshThreshold = refreshThreshold
    }
    if !payload.additionalParameters.isEmpty {
      config.additionalParameters = payload.additionalParameters
    }
    if let logger {
      config.logger = logger
    }

    if let openId = payload.openId, let openIdConfig = buildOpenIdConfiguration(openId) {
      config.openId = openIdConfig
    }

    if let storage = buildStorageDelegate(payload.storageId, queueKey: queueKey) {
      config.storage = storage
    }

    return config
  }

  /// Build a native OIDC web client from the payload.
  ///
  /// - Parameter payload: Parsed JS client configuration payload.
  /// - Parameter logger: Optional native logger resolved from Core registry.
  /// - Returns: Native OIDC web client.
  static func buildWebClient(
    _ payload: OidcClientPayload,
    logger: Logger?,
    queueKey: DispatchSpecificKey<Void>? = nil
  ) -> OidcWeb {
    return OidcWeb.createOidcWeb { config in
      if let logger {
        config.logger = logger
      }
      if let browserType = mapBrowserType(payload.browserType) {
        config.browserType = browserType
      }
      if let browserMode = mapBrowserMode(payload.browserMode) {
        config.browserMode = browserMode
      }
      config.module(OidcModule.config) { oidc in
        oidc.clientId = payload.clientId
        oidc.discoveryEndpoint = payload.discoveryEndpoint ?? ""
        oidc.redirectUri = payload.redirectUri
        oidc.scopes = Set(payload.scopes)
        oidc.acrValues = payload.acrValues
        oidc.state = payload.state
        oidc.nonce = payload.nonce
        oidc.uiLocales = payload.uiLocales
        oidc.loginHint = payload.loginHint
        oidc.display = payload.display
        oidc.prompt = payload.prompt
        if let refreshThreshold = payload.refreshThreshold {
          oidc.refreshThreshold = refreshThreshold
        }
        if !payload.additionalParameters.isEmpty {
          oidc.additionalParameters = payload.additionalParameters
        }
        if let openId = payload.openId, let openIdConfig = buildOpenIdConfiguration(openId) {
          oidc.openId = openIdConfig
        }
        if let storage = buildStorageDelegate(payload.storageId, queueKey: queueKey) {
          oidc.storage = storage
        }
      }
    }
  }

  /// Build an OIDC storage delegate if a storage id is provided.
  ///
  /// - Parameter storageId: Storage identifier from JS.
  /// - Returns: Storage delegate or nil when not configured.
  private static func buildStorageDelegate(
    _ storageId: String?,
    queueKey: DispatchSpecificKey<Void>?
  ) -> StorageDelegate<Token>? {
    guard let storageId, !storageId.isEmpty else {
      return nil
    }
    guard let queueKey,
          let config = resolveStorageConfigFromCoreSync(storageId, queueKey: queueKey) else {
      return nil
    }
    let account = config.account ?? "ACCESS_TOKEN_STORAGE"
    let encryptorEnabled = config.encryptor ?? true
    let encryptor: Encryptor = {
      if encryptorEnabled, let secured = SecuredKeyEncryptor() {
        return secured
      }
      return NoEncryptor()
    }()
    return KeychainStorage<Token>(
      account: account,
      encryptor: encryptor,
      cacheStrategy: parseCacheStrategy(cacheable: config.cacheable, rawStrategy: nil)
    )
  }

  /// Resolves storage configuration from Core OIDC registry by id.
  ///
  /// - Parameters:
  ///   - storageId: Storage handle identifier from JS.
  ///   - queueKey: Specific key for the queue that is allowed to block.
  /// - Returns: Storage handle contract payload, or nil.
  private static func resolveStorageConfigFromCoreSync(
    _ storageId: String,
    queueKey: DispatchSpecificKey<Void>
  ) -> StorageConfigHandleContract? {
    return RegistrySync.resolveSync(
      storageId,
      registry: CoreRuntime.oidcStorageConfigRegistry,
      queueKey: queueKey,
      context: "OidcClientFactory.resolveStorageConfigFromCoreSync"
    ) as? StorageConfigHandleContract
  }

  /// Maps storage config payload values to PingStorage cache strategy.
  ///
  /// Supports both the current `cacheable` boolean and future `cacheStrategy` string.
  ///
  /// - Parameter config: Registered storage configuration payload.
  /// - Returns: Native cache strategy value.
  private static func parseCacheStrategy(cacheable: Bool?, rawStrategy: String?) -> CacheStrategy {
    if let rawStrategy = rawStrategy?.lowercased() {
      switch rawStrategy {
      case "cache":
        return .CACHE
      case "cache_on_failure":
        return .CACHE_ON_FAILURE
      case "no_cache":
        return .NO_CACHE
      default:
        break
      }
    }

    if let cacheable {
      return cacheable ? .CACHE_ON_FAILURE : .NO_CACHE
    }

    return .NO_CACHE
  }

  /// Build OpenID configuration using the Codable initializer.
  ///
  /// - Parameter openId: Parsed OpenID override payload.
  /// - Returns: OpenID configuration or nil when serialization fails.
  private static func buildOpenIdConfiguration(_ openId: OpenIdPayload) -> OpenIdConfiguration? {
    var payload: [String: Any] = [
      "authorization_endpoint": openId.authorizationEndpoint,
      "token_endpoint": openId.tokenEndpoint,
      "userinfo_endpoint": openId.userinfoEndpoint,
      "end_session_endpoint": openId.endSessionEndpoint ?? "",
      "revocation_endpoint": openId.revocationEndpoint ?? ""
    ]
    if let pingEnd = openId.pingEndIdpSessionEndpoint {
      payload["ping_end_idp_session_endpoint"] = pingEnd
    }
    guard let data = try? JSONSerialization.data(withJSONObject: payload, options: []) else {
      return nil
    }
    return try? JSONDecoder().decode(OpenIdConfiguration.self, from: data)
  }

  /// Map JS string values to PingBrowser BrowserType.
  ///
  /// - Parameter value: JS browser type string.
  /// - Returns: Parsed BrowserType or nil when unsupported.
  private static func mapBrowserType(_ value: String?) -> BrowserType? {
    guard let value = value?.lowercased() else {
      return nil
    }
    switch value {
    case "authsession":
      return .authSession
    case "nativebrowserapp":
      return .nativeBrowserApp
    case "sfviewcontroller":
      return .sfViewController
    case "ephemeralauthsession":
      return .ephemeralAuthSession
    default:
      return nil
    }
  }

  /// Map JS string values to PingBrowser BrowserMode.
  ///
  /// - Parameter value: JS browser mode string.
  /// - Returns: Parsed BrowserMode or nil when unsupported.
  private static func mapBrowserMode(_ value: String?) -> BrowserMode? {
    guard let value = value?.lowercased() else {
      return nil
    }
    switch value {
    case "login":
      return .login
    case "logout":
      return .logout
    case "custom":
      return .custom
    default:
      return nil
    }
  }
}
